go/nbs/s3_table_persister.go

```diff
@@ -184,3 +194,180 @@ func (s partsByPartNum) Less(i, j int) bool {
 func (s partsByPartNum) Swap(i, j int) {
 	s[i], s[j] = s[j], s[i]
 }
+
+func (s3p s3TablePersister) CompactAll(sources chunkSources) chunkSource {
+	plan := planCompaction(sources)
+	if plan.chunkCount == 0 {
+		return emptyChunkSource{}
+	}
+	t1 := time.Now()
+	name := nameFromSuffixes(plan.suffixes())
+	s3p.multipartCopyUpload(plan, name.String())
+	verbose.Log("Compacted table of %d Kb in %s", plan.totalCompressedData/1024, time.Since(t1))
+
+	return s3p.newReaderFromIdxData(plan.mergedIndex, name)
+}
+
+func (s3p s3TablePersister) multipartCopyUpload(plan compactionPlan, key string) {
+	uploadID := s3p.startMultipartUpload(key)
+	multipartUpload, err := s3p.uploadPartsCopy(plan, key, uploadID)
+	if err != nil {
+		s3p.abortMultipartUpload(key, uploadID)
+		d.PanicIfError(err) // TODO: Better error handling here
+	}
+	s3p.completeMultipartUpload(key, uploadID, multipartUpload)
+}
+
+func (s3p s3TablePersister) uploadPartsCopy(plan compactionPlan, key, uploadID string) (*s3.CompletedMultipartUpload, error) {
```

Comment:
This function is pretty dense. How about some comments that help the reader grok what's going on.